File: design-pattern-jvm/app/src/main/java/design/pattern/java/patterns/structural/proxy/ServiceProxy.java
package design.pattern.java.patterns.structural.proxy;

public class ServiceProxy implements Service {
  private final Service service;
  private final ServiceCacheFactory cache;

  public ServiceProxy(Service service) {
    this.service = service;
    this.cache = ServiceCacheFactory.create();
  }

  @Override
  public String addPrefix(String word) {
    if (this.cache.isCached(word)) {
      return (String) this.cache.getInCache(word);
    }
    return this.cache.saveInCache(word, this.service.addPrefix(word));
  }

}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/structural/proxy/ServiceProxy.kt
package design.pattern.kotlin.patterns.structural.proxy

class ServiceProxy(private val service: Service) : Service {
    private val cache: ServiceCacheFactory? = ServiceCacheFactory.create()
    override fun addPrefix(word: String): String {
        return if (cache == null) "" else if (cache.isCached(word)) cache
            .getInCache(word) as String else cache.saveInCache(
            word,
            service.addPrefix(word)
        )
    }

}